From cc5caeab1ab8f5c3dc71594bdfd3281c39566754 Mon Sep 17 00:00:00 2001
From: Mathis MARION <mathis.marion@grenoble-inp.org>
Date: Thu, 29 Jul 2021 16:58:07 +0200
Subject: [PATCH 15/18] [fmc-adc-100m14b4cha-sw] Kernel 5.10 table alloc

The signature for `__sg_alloc_table_from_pages` changed in
kernel 5.10 and onwards.
Two agruments are added which can be passed as NULL and 0
for the same behavior as before, except for error checking,
which was addressed in this patch.
---
 patches/fmc-adc-100m14b4cha-sw.diff | 40 ++++++++++++++++++++++++++++-
 1 file changed, 39 insertions(+), 1 deletion(-)

diff --git a/patches/fmc-adc-100m14b4cha-sw.diff b/patches/fmc-adc-100m14b4cha-sw.diff
index c184ac0..f515366 100644
--- a/patches/fmc-adc-100m14b4cha-sw.diff
+++ b/patches/fmc-adc-100m14b4cha-sw.diff
@@ -1,5 +1,5 @@
 diff --git a/kernel/fa-core.c b/kernel/fa-core.c
-index 6e1eb6d..8a21683 100644
+index 6e1eb6d..532b0a8 100644
 --- a/kernel/fa-core.c
 +++ b/kernel/fa-core.c
 @@ -10,6 +10,7 @@
@@ -10,6 +10,44 @@ index 6e1eb6d..8a21683 100644
  #include "linux/ipmi-fru.h"
  #include "linux/fmc.h"
  
+@@ -38,6 +39,28 @@ static const int zfad_hw_range[] = {
+ struct workqueue_struct *fa_workqueue;
+ 
+ 
++static int fa_sg_alloc_table_from_pages(struct sg_table *sgt,
++						struct page **pages,
++						unsigned int n_pages,
++						unsigned int offset,
++						unsigned long size,
++						unsigned int max_segment,
++						gfp_t gfp_mask)
++{
++#if KERNEL_VERSION(5, 10, 0) <= LINUX_VERSION_CODE
++	struct scatterlist *sg;
++	sg =  __sg_alloc_table_from_pages(sgt, pages, n_pages, offset, size,
++			max_segment, NULL, 0, gfp_mask);
++	if (IS_ERR(sg))
++		return PTR_ERR(sg);
++	else
++		return 0;
++#else
++	return __sg_alloc_table_from_pages(sgt, pages, n_pages, offset, size,
++			max_segment, gfp_mask);
++#endif
++}
++
+ /**
+  * Description:
+  *    The version from the Linux kernel automatically squash contiguous pages.
+@@ -558,7 +581,7 @@ int fa_probe(struct platform_device *pdev)
+ 	case ADC_VER_SPEC:
+ 		fa->memops.read = ioread32;
+ 		fa->memops.write = iowrite32;
+-		fa->sg_alloc_table_from_pages = __sg_alloc_table_from_pages;
++		fa->sg_alloc_table_from_pages = fa_sg_alloc_table_from_pages;
+ 		break;
+ 	case ADC_VER_SVEC:
+ 		fa->memops.read = ioread32be;
 diff --git a/kernel/fa-debug.c b/kernel/fa-debug.c
 index 7478641..3e99eb6 100644
 --- a/kernel/fa-debug.c
-- 
2.27.0

